In [2]:
import pandas as pd
df = pd.read_csv(r'American_house_prices_dataset.csv')
df.head()

Unnamed: 0.1  Unnamed: 0        zpid rawHomeStatusCd  \
0             0           0  59141778.0         Pending   
1             1           1  59172422.0    RecentlySold   
2             2           2         NaN             NaN   
3             3           3  59109603.0    RecentlySold   
4             4           4  59110140.0    RecentlySold   

  marketingStatusSimplifiedCd  \
0          Non Owner Occupied   
1                RecentChange   
2                         NaN   
3                RecentChange   
4                RecentChange   

                                              imgSrc hasImage  \
0  https://photos.zillowstatic.com/fp/8e78d9c3127...     True   
1  https://photos.zillowstatic.com/fp/17fa75ec2c3...     True   
2  https://photos.zillowstatic.com/fp/db4bb4b2af8...     True   
3  https://photos.zillowstatic.com/fp/72bb4e8b21a...     True   
4  https://photos.zillowstatic.com/fp/831ef787fbd...     True   

                                           detailUrl statusType statusText  \
0  /homedetails/9-Robken-Rd-Roslindale-MA-02131/5...       SOLD       Sold   
1  /homedetails/130-Appleton-St-APT-2A-Boston-MA-...       SOLD       Sold   
2    /b/70-leo-m-birmingham-pkwy-brighton-ma-5j3ZG6/       SOLD   For Rent   
3  /homedetails/64-Bakersfield-St-Dorchester-MA-0...       SOLD       Sold   
4  /homedetails/123-Norwell-St-Dorchester-MA-0212...       SOLD       Sold   

   ... info2String badgeInfo.type badgeInfo.text  \
0  ...         NaN            NaN            NaN   
1  ...         NaN            NaN            NaN   
2  ...         NaN            NaN            NaN   
3  ...         NaN            NaN            NaN   
4  ...         NaN            NaN            NaN   

   hdpData.homeInfo.listing_sub_type.is_forAuction  isR4R  \
0                                              NaN    NaN   
1                                              NaN    NaN   
2                                              NaN    NaN   
3                                              NaN    NaN   
4                                              NaN    NaN   

   hdpData.homeInfo.rentalMarketingSubType communityName isCdpResult  \
0                                      NaN           NaN         NaN   
1                                      NaN           NaN         NaN   
2                                      NaN           NaN         NaN   
3                                      NaN           NaN         NaN   
4                                      NaN           NaN         NaN   

  isComingSoonCommunity style  
0                   NaN   NaN  
1                   NaN   NaN  
2                   NaN   NaN  
3                   NaN   NaN  
4                   NaN   NaN  

[5 rows x 111 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7503 entries, 0 to 7502
Columns: 111 entries, Unnamed: 0.1 to style
dtypes: bool(1), float64(32), int64(2), object(76)
memory usage: 6.3+ MB


In [4]:
df.drop(df.iloc[:,[0,1]],axis=1,inplace=True)

In [5]:
null_=df.isnull().sum()
print(null_.shape[0])
#print(df.shape[1])

109


In [6]:
null_index=null_[null_.iloc[:]>0.20*df.shape[0]].index
print(len(null_index))

50


In [7]:
df.drop(columns=null_index,axis=1,inplace=True)

In [8]:
df.head()

zpid rawHomeStatusCd marketingStatusSimplifiedCd  \
0  59141778.0         Pending          Non Owner Occupied   
1  59172422.0    RecentlySold                RecentChange   
2         NaN             NaN                         NaN   
3  59109603.0    RecentlySold                RecentChange   
4  59110140.0    RecentlySold                RecentChange   

                                              imgSrc hasImage  \
0  https://photos.zillowstatic.com/fp/8e78d9c3127...     True   
1  https://photos.zillowstatic.com/fp/17fa75ec2c3...     True   
2  https://photos.zillowstatic.com/fp/db4bb4b2af8...     True   
3  https://photos.zillowstatic.com/fp/72bb4e8b21a...     True   
4  https://photos.zillowstatic.com/fp/831ef787fbd...     True   

                                           detailUrl statusType statusText  \
0  /homedetails/9-Robken-Rd-Roslindale-MA-02131/5...       SOLD       Sold   
1  /homedetails/130-Appleton-St-APT-2A-Boston-MA-...       SOLD       Sold   
2    /b/70-leo-m-birmingham-pkwy-brighton-ma-5j3ZG6/       SOLD   For Rent   
3  /homedetails/64-Bakersfield-St-Dorchester-MA-0...       SOLD       Sold   
4  /homedetails/123-Norwell-St-Dorchester-MA-0212...       SOLD       Sold   

      price priceLabel  ... hdpData.homeInfo.isUnmappable  \
0  $819,500      $820K  ...                         False   
1  $880,000      $880K  ...                         False   
2       NaN        NaN  ...                           NaN   
3    $1.25M     $1.25M  ...                         False   
4  $805,000      $805K  ...                         False   

   hdpData.homeInfo.isPreforeclosureAuction  hdpData.homeInfo.priceForHDP  \
0                                     False                      819500.0   
1                                     False                      880000.0   
2                                       NaN                           NaN   
3                                     False                     1250000.0   
4                                     False                      805000.0   

   hdpData.homeInfo.isNonOwnerOccupied hdpData.homeInfo.isPremierBuilder  \
0                                 True                             False   
1                                 True                             False   
2                                  NaN                               NaN   
3                                 True                             False   
4                                 True                             False   

  hdpData.homeInfo.isZillowOwned hdpData.homeInfo.currency  \
0                          False                       USD   
1                          False                       USD   
2                            NaN                       NaN   
3                          False                       USD   
4                          False                       USD   

  hdpData.homeInfo.country hdpData.homeInfo.isShowcaseListing  \
0                      USA                              False   
1                      USA                              False   
2                      NaN                                NaN   
3                      USA                              False   
4                      USA                              False   

   hdpData.homeInfo.homeStatusForHDP  
0                                NaN  
1                      RECENTLY_SOLD  
2                                NaN  
3                      RECENTLY_SOLD  
4                      RECENTLY_SOLD  

[5 rows x 59 columns]

In [53]:
# df.to_csv("american_hpd.csv",index=False)

In [9]:
indices_to_remove=df[df['hdpData.homeInfo.priceForHDP'].isnull()==True].index
print(indices_to_remove)


Index([   2,    8,   10,   11,   13,   20,   22,   25,   26,   27,
       ...
       7464, 7466, 7467, 7468, 7475, 7486, 7487, 7489, 7491, 7502],
      dtype='int64', length=1297)


In [10]:
# Assuming 'indices_to_remove' is a list of indices you want to remove
df.drop(index=indices_to_remove, inplace=True)


In [11]:
df.shape[0]

6206

1.hdpData.homeInfo.country-since every country is usa

2.hdpData.homeInfo.currency- obivously dollars

3.hdpData.homeInfo.priceForHDP and (price) - both has same but first one has int data type directly but not second one

4.priceLabel - price is just written in K's

5.address - address has been divided into columns already we don't need it

6.detailUrl - it's a link

7.imgSrc - no significance

8.hdpData.homeInfo.zpid, 9.zpid - useless

10.shouldShowZestimateAsPrice - 7502 are false and blanks / 1 true value

has3DModel - useless may be usefull, more false and less true 

hasVide0 - useless may be usefull,  more false and less true 

11.isHomeRec - only false and null (home recommendation) so it can have same reason as 15(visited)

12.isShowcaseListing-redundant with hdpData.homeInfo.isShowcaseListing

13.isFavorite-useless

14 ( isFeaturedListing ) - it has many false and few true's

15 visited - if nobody visits they gave false so they know who visited and who are not so null values means they didnot get the data

16.info6String - it has more 6500 blanks and 600 rows has many names

17,18. hdpData.homeInfo.latitude, hdpData.homeInfo.longitude - we lattitude and longitude attributes and these two columns has null values but "lattitude" and "longitude" doesnt have null values

hdpData.homeInfo.dateSold - instead of having date it has some weird number we need to check this 

19,20,21. hdpData.homeInfo.bathrooms,hdpData.homeInfo.bedrooms,hdpData.homeInfo.livingArea - redundant

hdpData.homeInfo.daysOnZillow - it has just -1 and null values but may be important mostly not

hdpData.homeInfo.isFeatured - it has 7481 rows with just false and null values, i don't think it will be usefull

22,23,24. ( hdpData.homeInfo.shouldHighlight, hdpData.homeInfo.isUnmappable, hdpData.homeInfo.isPreforeclosureAuction ) - it has just false and null values

25.hdpData.homeInfo.priceForHDP - redundant

hdpData.homeInfo.isNonOwnerOccupied - it has only true and null values

hdpData.homeInfo.isPremierBuilder - it has only false and null values but with above columns it perfectly correlated , so we need to think about this

hdpData.homeInfo.isZillowOwned - like isPremierBuilder is also correlated with above two 
i.e, all three are correlated either positively or negatively

hdpData.homeInfo.isShowcaseListing - all four are correlated we need to think

hdpData.homeInfo.streetAddress - 6202 unique values out of 6206 so no use

hdpData.homeInfo.zipcode - zipcode has many unique city codes and these any won't contribute info for information 

hdpData.homeInfo.homeStatus - "recently sold" and "pending" are column values this is dataset of sold houses and many columns has recently sold 
 
hdpData.homeInfo.homeStatusForHDP - useless data like above but has 5 or 6 unique values more than 6000 are recently sold

hdpData.homeInfo.isUnmappable, hdpData.homeInfo.isPreforeclosureAuction - only false and blanks no use

rawHomeStatusCd,marketingStatusSimplifiedCd,statusType,statusText - no use

isUserConfirmedClaim-only false and null



In [12]:
# hdpData.homeInfo.streetAddress, hdpData.homeInfo.homeStatus

import numpy as np
y=df['hdpData.homeInfo.homeStatus']
x=df['hdpData.homeInfo.homeStatus'].value_counts().values
z=np.unique(x)
print(z)
print(y.nunique())
# print(arr[arr > constant])

[  18 6188]
2


In [13]:
columns_1=['hdpData.homeInfo.country','hdpData.homeInfo.currency'] # obivious
columns_2=['price','priceLabel','address','isShowcaseListing','hdpData.homeInfo.bathrooms',
           'hdpData.homeInfo.bedrooms','hdpData.homeInfo.livingArea','hdpData.homeInfo.latitude','hdpData.homeInfo.longitude'] # redundant
# we can write code and remove columns may be using correlation or covariance bcoz these values corr for redundant columns is 1 and covariance is 0 
# but few cols may have entirely different col values but may have good correlation then we may loose a col
# but here these cols same values with different col names
columns_3=['zpid','hdpData.homeInfo.zpid','detailUrl','imgSrc','isFavorite','visited','hdpData.homeInfo.homeStatus','hdpData.homeInfo.homeStatusForHDP',
           'hdpData.homeInfo.daysOnZillow','hdpData.homeInfo.dateSold','isHomeRec','hdpData.homeInfo.streetAddress','hdpData.homeInfo.zipcode'] # useless
# cannot be coded for these cols , removing these cols manually is better  
columns_4=['shouldShowZestimateAsPrice','isFeaturedListing','hdpData.homeInfo.isFeatured','hdpData.homeInfo.isNonOwnerOccupied',
           'hdpData.homeInfo.isPremierBuilder','hdpData.homeInfo.isZillowOwned','has3DModel','hasVide0',
           'hdpData.homeInfo.isUnmappable','hdpData.homeInfo.isPreforeclosureAuction'] # more false and less true and may have null values/ too many false's
# we use for loop and remove cols for either col_value percentage is too high
columns_5=['info6String'] # non-sense

print(len(columns_1))
print(len(columns_2))
print(len(columns_3))
print(len(columns_4))
print(len(columns_5))
print(len(columns_1)+
len(columns_2)+
len(columns_3)+
len(columns_4)+
len(columns_5))
print(59-35)

2
9
13
10
1
35
24


In [17]:
cols_to_del = df.loc[:, df.nunique() == 1].columns
print(cols_to_del)
len(cols_to_del)

Index(['hasImage', 'statusType', 'isUserClaimingOwner', 'isUserConfirmedClaim',
       'isHomeRec', 'isShowcaseListing', 'isFavorite', 'visited',
       'hdpData.homeInfo.daysOnZillow', 'hdpData.homeInfo.shouldHighlight',
       'hdpData.homeInfo.isUnmappable',
       'hdpData.homeInfo.isPreforeclosureAuction',
       'hdpData.homeInfo.isNonOwnerOccupied', 'hdpData.homeInfo.isZillowOwned',
       'hdpData.homeInfo.currency', 'hdpData.homeInfo.country',
       'hdpData.homeInfo.isShowcaseListing'],
      dtype='object')


17

In [19]:
# removing cols which have one value that means no variance so no info
## df.drop(columns=cols_to_del,axis=1,inplace=True)
df.shape


(6206, 42)

In [23]:
# deleting obivious and redundant data
# columns_1 cols have been removed in above code so we don't need to include columns_1 list and more col isShowcaseListing
## columns_2.remove('isShowcaseListing')
## df.drop(columns=columns_2,axis=1,inplace=True)
df.shape

(6206, 34)

In [27]:
# dropping useless cols which are searched manually
# alrdy_rm_cols=['isFavorite', 'visited', 'hdpData.homeInfo.daysOnZillow', 'isHomeRec'] 
# for x in alrdy_rm_cols:
#     columns_3.remove(x)
## df.drop(columns=columns_3,axis=1,inplace=True)
df.shape

(6206, 25)

In [30]:
# alrdy_rm_cols=['hdpData.homeInfo.isNonOwnerOccupied','hdpData.homeInfo.isZillowOwned', 'hasVide0',
#                   'hdpData.homeInfo.isUnmappable', 'hdpData.homeInfo.isPreforeclosureAuction']
# for x in alrdy_rm_cols:
#     columns_4.remove(x)
## df.drop(columns=columns_4,axis=1,inplace=True)
df.shape

(6206, 20)

In [31]:
df.columns

Index(['rawHomeStatusCd', 'marketingStatusSimplifiedCd', 'statusText', 'beds',
       'baths', 'area', 'pgapt', 'sgapt', 'hasVideo',
       'hasAdditionalAttributions', 'latLong.latitude', 'latLong.longitude',
       'variableData.type', 'variableData.text', 'hdpData.homeInfo.city',
       'hdpData.homeInfo.state', 'hdpData.homeInfo.price',
       'hdpData.homeInfo.homeType', 'hdpData.homeInfo.rentZestimate',
       'hdpData.homeInfo.priceForHDP'],
      dtype='object')

In [40]:
# rawHomeStatusCd,marketingStatusSimplifiedCd,statusText
# hdpData.homeInfo.priceForHDP is redundant we didn't check
## df.drop(columns=['hdpData.homeInfo.priceForHDP','rawHomeStatusCd','marketingStatusSimplifiedCd','statusText','pgapt'],axis=1,inplace=True)
df.shape

(6206, 15)

In [41]:
df.columns

Index(['beds', 'baths', 'area', 'sgapt', 'hasVideo',
       'hasAdditionalAttributions', 'latLong.latitude', 'latLong.longitude',
       'variableData.type', 'variableData.text', 'hdpData.homeInfo.city',
       'hdpData.homeInfo.state', 'hdpData.homeInfo.price',
       'hdpData.homeInfo.homeType', 'hdpData.homeInfo.rentZestimate'],
      dtype='object')

In [57]:
df['variableData.type'].value_counts()

variableData.type
RECENTLY_SOLD    6123
OPEN_HOUSE         10
Name: count, dtype: int64

In [52]:
x=df['sgapt'].value_counts()/(6206)
x.values
for i in range(len(x.values)):
    if x[i]>0.80:
        df.drop(columns=)

array([0.60602643, 0.39397357])

In [ ]:
# columns_1=['hdpData.homeInfo.country','hdpData.homeInfo.currency'] # obivious
# columns_2=['price','priceLabel','address','isShowcaseListing','hdpData.homeInfo.bathrooms',
#            'hdpData.homeInfo.bedrooms','hdpData.homeInfo.livingArea','hdpData.homeInfo.latitude','hdpData.homeInfo.longitude'] # redundant
# columns_3=['zpid','hdpData.homeInfo.zpid','detailUrl','imgSrc','isFavorite','visited',
#            'hdpData.homeInfo.daysOnZillow','hdpData.homeInfo.dateSold','isHomeRec'] # useless
# columns_4=['shouldShowZestimateAsPrice','isFeaturedListing','hdpData.homeInfo.isFeatured','hdpData.homeInfo.isNonOwnerOccupied',
#            'hdpData.homeInfo.isPremierBuilder','hdpData.homeInfo.isZillowOwned'] # more false and less true and may have null values/ too many false's
# columns_5=['info6String'] # non-sense